flow/display_list.h

```diff
@@ -0,0 +1,427 @@
+  V(DrawPath)                       \
+                                    \
+  V(DrawPoints)                     \
+  V(DrawLines)                      \
+  V(DrawPolygon)                    \
+  V(DrawVertices)                   \
+                                    \
+  V(DrawImage)                      \
+  V(DrawImageRect)                  \
+  V(DrawImageNine)                  \
+  V(DrawImageLattice)               \
+  V(DrawAtlas)                      \
+  V(DrawAtlasColored)               \
+  V(DrawAtlasCulled)                \
+  V(DrawAtlasColoredCulled)         \
+                                    \
+  V(DrawSkPicture)                  \
+  V(DrawDisplayList)                \
+  V(DrawTextBlob)                   \
+  /* V(DrawShadowRec) */            \
+                                    \
+  V(DrawShadow)
+
+#define DL_OP_TO_ENUM_VALUE(name) name,
+enum DisplayListOpType { FOR_EACH_DISPLAY_LIST_OP(DL_OP_TO_ENUM_VALUE) };
+#undef DL_OP_TO_ENUM_VALUE
+
+class Dispatcher;
+class DisplayListBuilder;
+
+// The base class that contains a sequence of rendering operations
+// for dispatch to a Dispatcher. These objects must be instantiated
+// through an instance of DisplayListBuilder::build().
+class DisplayList : public SkRefCnt {
+ public:
+  static const SkSamplingOptions NearestSampling;
+  static const SkSamplingOptions LinearSampling;
+  static const SkSamplingOptions MipmapSampling;
+  static const SkSamplingOptions CubicSampling;
+
+  DisplayList()
+      : ptr_(nullptr),
+        used_(0),
+        opCount_(0),
+        uniqueID_(0),
+        bounds_({0, 0, 0, 0}),
+        boundsCull_({0, 0, 0, 0}) {}
+
+  ~DisplayList();
+
+  void dispatch(Dispatcher& ctx) const { dispatch(ctx, ptr_, ptr_ + used_); }
+
+  void renderTo(SkCanvas* canvas) const;
+
+  size_t bytes() const { return used_; }
+  int opCount() const { return opCount_; }
+  uint32_t uniqueID() const { return uniqueID_; }
+
+  const SkRect& bounds() {
+    if (bounds_.width() < 0.0) {
+      // computeBounds() will leave the variable with a
+      // non-negative width and height
+      computeBounds();
+    }
+    return bounds_;
+  }
+
+  bool equals(const DisplayList& other) const;
+
+ private:
+  DisplayList(uint8_t* ptr, size_t used, int opCount, const SkRect& cullRect);
+
+  uint8_t* ptr_;
+  size_t used_;
+  int opCount_;
+
+  uint32_t uniqueID_;
+  SkRect bounds_;
+
+  // Only used for drawPaint() and drawColor()
+  SkRect boundsCull_;
+
+  void computeBounds();
+  void dispatch(Dispatcher& ctx, uint8_t* ptr, uint8_t* end) const;
+
+  friend class DisplayListBuilder;
+};
+
+// The pure virtual interface for interacting with a display list.
+// This interface represents the methods used to build a list
+// through the DisplayListBuilder and also the methods that will
+// be invoked through the DisplayList::dispatch() method.
+class Dispatcher {
+ public:
+  // MaxDrawPointsCount * sizeof(SkPoint) must be less than 1 << 32
```

Comment:
```
  static_assert(kMaxDrawPointsCount * sizeof(SkPoint) < 1 << 32);
  static constexpr int kMaxDrawPointsCount = ...;
```

Reply:
It actually needs to be much smaller. At the time I was looking at the "count" parameter being an int, but we encode the size of the op in a 24-bit field, so we can only support enough points to fit into the pod data area.